src/tail/session.rs

```diff
@@ -36,41 +44,66 @@ impl Session {
                 let tail_id = success.result.id;
 
                 // Loop indefinitely to send "heartbeat" to API and keep log streaming alive.
+                // This should loop forever until SIGINT is issued or Wrangler process is killed
+                // through other means.
+                let duration = Duration::from_millis(1000 * KEEP_ALIVE_INTERVAL);
+                let mut delay = time::delay_for(duration);
+
                 loop {
-                    thread::sleep(Duration::from_secs(60));
-                    let heartbeat_result = send_heartbeat(target, &client, &tail_id).await;
-                    if heartbeat_result.is_err() {
-                        return heartbeat_result;
+                    match rx.try_recv() {
+                        Err(TryRecvError::Empty) => {
+                            if delay.is_elapsed() {
+                                let heartbeat_result =
+                                    send_heartbeat(&target, &client, &tail_id).await;
+                                if heartbeat_result.is_err() {
+                                    return heartbeat_result;
+                                }
+                                delay = time::delay_for(duration);
+                            }
+                        }
+                        _ => {
+                            return Ok(());
+                        }
                     }
-                    // This should loop forever until SIGINT is issued or Wrangler process is killed
-                    // through other means.
                 }
             }
             Err(e) => failure::bail!(http::format_error(e, None)),
         }
     }
 }
 
-async fn get_tunnel_url() -> Result<String, failure::Error> {
+async fn get_tunnel_url(rx: &mut Receiver<()>) -> Result<String, failure::Error> {
     // regex for extracting url from cloudflared metrics port.
     let url_regex = Regex::new("userHostname=\"(https://[a-z.-]+)\"").unwrap();
 
     let mut attempt = 0;
+    let mut delay = time::delay_for(Duration::from_millis(attempt * attempt * 1000));
```

Comment:
worth adding a comment about exponential backoff

Reply:
also, this will be 0 always since you just set attempt to 0.

Reply:
the first one will, yes, but since rust doesn't have do while i have to set it as something, and on the first time i don't want a delay. in the while loop i then update to set it again each time the request fails.